### ISSUE TITLE
Various Game Boy problems

### ISSUE THREAD

**Issue by RushJet1** [shonumi/gbe-plus#51]
I don't know if these have since been fixed (as the latest Windows binary is over a year old) but I ran into a bunch of issues while trying to apply color to Mega Man V.

- The audio is extremely inaccurate, as if it's running at 30hz or something, it misses beats sometimes and all sound effects sound very sloppy.
- When "load custom graphics" is checked, specific areas of the game run very slowly, even if no custom graphics are loaded for those parts.  The stage select is the most constant one, makes the program take a full CPU core and it runs at maybe 5 FPS.  Unchecking that box fixes the issues temporarily until it is checked again.
- When dumping graphics, if I try to dump the same sprite more than once (if the original dump got messed up), it won't let me re-dump it.  This happens even if I delete the manifest and all dumped images.
- Mega Man V seems to swap OBJ sprites in and out from the BG, where they're stored as inverted-palette tiles.  Not sure how to colorize these as they take up the same addresses as the normal sprites apparently, and dumping them normally doesn't work.
- If I enable "auto dumping" i get thousands of files after maybe 30 seconds. It's dumping tons of duplicates, probably due to the way it swaps the OBJs in and out of the BG tileset.

**Comment by shonumi** [comment]
> The audio is extremely inaccurate, as if it's running at 30hz or something, it misses beats sometimes and all sound effects sound very sloppy.

Sounds like GBE+ isn't running fullspeed. Just tried editing the graphics for MMV and it ran fine on my end. This is likely due to several issues that made hashing graphics slow in 1.0, but have since been fixed (see below).

> When "load custom graphics" is checked, specific areas of the game run very slowly, even if no custom graphics are loaded for those parts. The stage select is the most constant one, makes the program take a full CPU core and it runs at maybe 5 FPS. Unchecking that box fixes the issues temporarily until it is checked again.

This happens because GBE+ generates new hashes every time it encounters new graphics and references those hashes to check if graphics need to be replaced. So the hash has to be created whenever VRAM changes. Unfortunately, there was no restriction on the size of the internal hash list in the 1.0 version, so it can get progressively longer, and thus take up more space and time to search, even when no graphics are replaced. This was fixed a long time ago, along with smarter conditions for generating hashes so that hashing takes place less frequently.

> When dumping graphics, if I try to dump the same sprite more than once (if the original dump got messed up), it won't let me re-dump it. This happens even if I delete the manifest and all dumped images.

Yup. Known issue I fixed sometime in April or March of this year. It made it into the 1.1 release, but there is no Windows 1.1 binary, but hopefully there will be for 1.2.

> Mega Man V seems to swap OBJ sprites in and out from the BG, where they're stored as inverted-palette tiles. Not sure how to colorize these as they take up the same addresses as the normal sprites apparently, and dumping them normally doesn't work.

The VRAM for OBJ and some parts of BG is actually the same. It's the same dot data, just with a different palette. The game doesn't actually use any of the OBJ tiles when drawing the BG. It's not necessary to dump the inverted OBJ tiles. They won't be used. Dumping them doesn't work because there was a bug that didn't account for palette swaps, but it's no longer present in 1.1.

> If I enable "auto dumping" i get thousands of files after maybe 30 seconds. It's dumping tons of duplicates, probably due to the way it swaps the OBJs in and out of the BG tileset.

Auto-dumping is intended only for limited circumstances, generally for small games (think Tetris, it's only 32KB) but currently, with the amount of new tools that were added in 1.1, it's functionally irrelevant. There's a warning in the manual; it's only for people who know what they're doing. It dumps **everything**. The feature was slightly improved thanks to better conditions that determine when new hashes need to be generated, so it doesn't pick up nearly as many garbage or unused tiles, but it's still an advanced option most people shouldn't touch. It's there as a "just in case", should it ever prove useful. However, it is not the recommended way to dump tiles in the latest revisions.

I'd advise you to wait for 1.2 to come out, or try compiling it on Linux if you want a quick preview. I'm still working on issues that prevented newer revisions from compiling on Windows (thus no binary for 1.1), but most of what you describe are problems specific to 1.0. I'll mark it as invalid but leave this open for further discussion. I'll also double-check everything later tonight just to make sure the issues really do no longer exist in the current revisions.

**Comment by RushJet1** [comment]
Yeah I don't really have time to mess with compiling it myself :( 

Anyway, I recorded a couple videos to show the sound effect difference (there are others, nearly every sound effect sounds wrong). The game seems to be running at full speed, the program takes like 1-4% of my CPU at any given time except for the stage select, which you can see is super slow until i temporarily disabled custom graphics.

GBE+ sound output: https://www.youtube.com/watch?v=V6YEaQwqo3I
BGB sound output (accurate): https://www.youtube.com/watch?v=f_atgAYMI44

**Comment by shonumi** [comment]
> Yeah I don't really have time to mess with compiling it myself :(

That's fine. Just know that 1.2 isn't due until next year in April.\

> Anyway, I recorded a couple videos to show the sound effect difference (there are others, nearly every sound effect sounds wrong). The game seems to be running at full speed, the program takes like 1-4% of my CPU at any given time except for the stage select, which you can see is super slow until i temporarily disabled custom graphics.

Thanks for taking the time to make those vids. I went and triple-checked, and the sound quality in the recent revisions doesn't match the output from 1.0, with or without loading graphics. So whatever caused that to happen has been fixed.

There are some areas where DMG/GBC sound can be improved in GBE+ (stereo support, better Channel 3 support, faster updates for higher quality sound, and some sort of low-pass filter) but I played around a bunch with MMV and no missing notes, and the sound effects more or less match BGB and VBA-M (plus or minus some filtering).

So I'd recommend you sit tight until 1.2 comes out. CGFX went through **a lot** of UI improvements (with more on the way). IMO, I can't imagine anyone doing any serious work replacing graphics on 1.0; it's just too cumbersome, especially when it comes to sprites.

**Comment by RushJet1** [comment]
Ah, April... I was hoping to have a solution to colorizing MMV by the end of this year (I tried "colorit" but it just allows you to change palettes, which is much more restrictive and cumbersome).  What do I need to compile this in Linux? I haven't compiled anything in Linux since 2008 though so it might be more work than I want to do....

**Comment by shonumi** [comment]
Just SDL2, OpenGL 3.3+ (any version of Mesa from the past, 10 years should do, just about), CMake, and Qt. I have not upgraded GBE+ to Qt5 yet. It works fine on Qt4, but there may be some minor issues when building with Qt5.

**Comment by RushJet1** [comment]
Got xubuntu running on a VM and have everything installed but cmake is whining about SDL2NET_LIBRARY. Its homepage basically says "contact your distro's info" but nothing says anything about SDL2NET... too tired to figure it out; I'll try again tomorrow.

**Comment by shonumi** [comment]
SDL2NET is an optional dependency. You only need it for netplay. You can disable it by changing Line 16 of the main CMakeLists.txt file. Just edit the "ON" part to say "OFF".

**Comment by RushJet1** [comment]
Set it to "OFF" and it still says 

![image](https://user-images.githubusercontent.com/4399475/32499116-a4214f9e-c39f-11e7-96e2-20c15980ba92.png)

**Comment by shonumi** [comment]
After editing the main CMakeLists.txt file, did you reset the CMake cache? Easiest way is to just delete the build folder and recreate it. Then run "cmake .." again.

**Comment by RushJet1** [comment]
OK so I compiled it and can run games, but can't output any bmp files.  I've chmodded the directories that are the output directories to allow read/write access but it just says that it can't output the bmp files and to check paths/permissions.

**Comment by shonumi** [comment]
Did you set the BG Dump and OBJ Dump paths? Be sure to double-check that. Those paths always point to GBE+'s data folder.

If you installed via  **make install**, the data folder is in your home directory. For you it would look like:

> /home/rushjet1/.gbe_plus/data/

You'll need to point GBE+ to that folder in the Paths option.

If you haven't installed it via **make install**, the GBE+ will look for the data folder in the same directory as the executable (only if a gbe.ini file is also present in the same directory as the executable). This is using GBE+'s "portable mode".

It was largely intended for Windows since you don't typically install emulators, rather, you just download them and extract the zip and run them. The portable mode also works on Linux, so long as you place the .ini file and the data folder alongside the executable. As a side note, the data folder should always have the shaders in there too if you use OpenGL.

So if you wanted to do portable mode:

1) Copy the data folder (https://github.com/shonumi/gbe-plus/tree/master/src/data) from the source and put it next to gbe_plus_qt (typically from the build folder). Optionally, make new folders for BG and OBJ dumps.

2) Copy the .ini file (https://github.com/shonumi/gbe-plus/blob/master/src/gbe.ini) from the source and put it next to gbe_plus_qt

3) Open up GBE+ and go to the Paths Settings, and point GBE+ to the folder you want to dump BG and OBJs.

And that should be it.

**Comment by RushJet1** [comment]
It was because I made a "dump" folder but not a "data" one, so the program was trying to find a folder that doesn't exist (without telling me though, I was wondering why it was saying "..\dump\BG" instead of just "dump\BG").  Putting "dump" inside "data" fixed the problem.

**Comment by RushJet1** [comment]
I'm able to colorize some sprites and it's much nicer than 1.0.

Some issues I've noticed:

-The slowdown for when you choose a level in Mega Man V still happens when "Load Custom Graphics" is checked.  Unchecking immediately fixes this, and I can just recheck in levels for testing purposes (or use savestates).

-Sound is still not perfect but it's decently better than 1.0

-Many meta OBJs in this game do not properly line up. For example, 

![mmjump](https://user-images.githubusercontent.com/4399475/32509693-16332802-c3bc-11e7-9d03-232b53b456e7.png)

This is still usable and works fine in-game, but it makes editing a little harder.  Maybe add an offset option per-column and/or per-row that would get added to the manifest file, like ignore x pixels before grabbing tiles, then offset the image by that many pixels for ease of use.  I'm sure other games probably have this issue too.

**Comment by shonumi** [comment]
For the graphics issues, can you send me your files? I suspect something went wrong in the dumping/editing process.

About the sound, I don't know what to tell you. It sounds fine on my end, and matches the output from VBA-M, with some differences in filtering. The only other issue I found was Mega Man's jump SFX ending sooner than it should, which I'm investigating currently. Other than that it sounds fine to me, even when using VBA-M side-by-side. Only thing I can think of is to set the sample rate to 48000Hz.

The performance issue might be hardware related. The hashing takes a lot of CPU power. I'm on a heavily OCed system, so I wouldn't notice it unless I downclocked. There may not be a way around it, since the hashing happens when a bunch of tiles in VRAM change, it could potentially bottleneck some systems.

**Comment by RushJet1** [comment]
Here are the edited files, https://www.dropbox.com/s/9jfv42h402dhphf/mmclr.zip?dl=0

I have a Core i7 4790K (4.4ghz single core) so I don't think even OCing would fix this problem for this system.

**Comment by shonumi** [comment]
Thanks for uploading them! I took a look, and it seems that the right side of your edit of Mega Man has a different color green than the left. It's hard to tell, but the hex values are different.

Now a couple of things, I went ahead and tested the stage select a couple of more times, and it seems to be doing something weird that other games don't. It's an edge-case where MMV updates the background palette every frame, but the palette is mostly the same. In GBE+, the emulator hashes every single BG tile whenever a palette changes to see if the new palette has any graphics that need to be replaced.

Normally, games only change the BG palette once every few frames when something happens (you enter a door and the screen fades to black, or some of the Pokemon attacks like Night Shade or whatever). So MMV causes all BG tiles (384) to be hashed at 60Hz, which is quite a lot. My i5-2500K at 4.0 GHz struggled, just like your computer. [That's all been fixed](https://github.com/shonumi/gbe-plus/commit/95c24ad3999e7ef56524cdc07c0ea2ffb5e9fd5c), so try updating the source code and see how that runs now.

About the OBJ meta-tiles, yeah, that's how Capcom arranged that stuff in VRAM. You can't piece it together correctly unless some of the tiles have X and Y offsets (which the Game Boy hardware took care of). Tthe files should work fine even if they look disjointed. They might look scrambled when you build your own sprite sheets through GBE+, but the data gets interpreted correctly, so don't be afraid to use them. The main issue is that it's not entirely friendly during the editing process.

It's tempting to add offsets to the manifest, but that would have to be a TODO for later. I'll definitely keep it in mind though.

**Comment by RushJet1** [comment]
Heh that would do it. The program I'm using is Pinta for these, and while it seems like a good simple solution so far, it seems to be resetting its palette constantly which is annoying.  I want to see if I can make custom palettes because that would be very useful for editing graphics.

I'll check out the changes later today when I have some more time.

**Comment by RushJet1** [comment]
OK The speed issue has been resolved.

One thing that I've noticed, when Mega Man charges his buster, it swaps palettes. ALL of the objects on screen also change palettes apparently when this happens, which means I have to dump 4 of each frame of animation for enemies and point them at the same graphic.  Is there a way you could make this easier to deal with? Enemies' palette swaps aren't visible to the player, just happens in game code.

**Comment by shonumi** [comment]
You would have to dump different versions of each one. There isn't a good way to do that just yet, but it's a problem I've been aware of for a while. It's a problem for games like Metroid 2 or Wario Land as well (when you pause each game, all pallettes change).

The solution I had in mind was to let people choose one palette to dump; other variations can be specified and custom graphics made for them. But only one set would need to be pulled from the game.

It's not hard conceptually, just non-trivial to implement. I'm looking into that for 1.2, hopefully getting it done next month rather than next year.

If it helps in the meantime, you can use the Frame Control options in GBE+. You can select inputs and advance frame-by-frame so you can capture different animations without opening and closing the CGFX menu constantly.

**Comment by RushJet1** [comment]
Yeah I've been using the frame control a lot- it's incredibly useful!  Basically I've just been using one image for all the palette changes and putting all the hashes in the manifest file manually.

**Comment by shonumi** [comment]
Okay, so let's turn this into a valid feature request for the CGFX​ UI. So this various improvements proposed:

1) OBJ Meta Tile offsets.

2) Replace graphics for multiple palette swaps without the need to dump separate graphics.

3) Allow user to replace, edit, and delete existing manifest entries.

An "Add to existing/replace existing" option sounds great. I've been planning this for a while, but on the Manifest tab, it would be very useful to delete or edit stuff right from the GUI, so you wouldn't have to mess around with the text file so much.

**Comment by RushJet1** [comment]
OK so I asked a hacker friend about the enemies changing palettes when Mega Man charges, and he says that there's nothing changing on the 2nd OBJ palette that they use, so he doesn't know why they'd change palettes.  I confirmed this with the debugger, when the custom graphics stop showing up, the palette stays the same, which is strange.

**Comment by shonumi** [comment]
Here's what I believe is going on. MMV uses two palettes, one for Mega Man and one for enemies. Mega Man's changes when he's charging, but the enemies stay the same. The thing is, GBE+ uses both palettes to make the hash for OBJs. So whenever OBP0 changes while Mega Man charges, enemy sprites technically generate a "new" hash. This was a crude way I came up with last year to deal with some issues regarding palette changes and hashing. The solution is easy, but the bad news is that it would invalidate all of your existing hashes in your manifest file. I'd have to change the hash format :(

However, the hash format change was bound to come sooner or later in order to take care of the second item on the list I made above. So it's unavoidable moving forward. The good news is that it shouldn't change again in the future.

About the BG screen, try adjusting the "Stop LCD rendering after line" option. Anything within 0 to 120 or something works I think. The DMG Mega Man games all do a bunch weird things with the BG and Window layer towards the bottom of the screen (to display the health bar and lives) but that only happens after the screen renders a certain line. I had to add that option specifically for the Mega Man games, and others like Aladdin.

**Comment by RushJet1** [comment]
Downloaded the latest source zip but can't build.

/home/rushjet1/gbeplus/5/gbe-plus-master/src/dmg/dmg07.cpp:158:84: error: ‘SDLNet_TCP_AddSocket’ was not declared in this scope
 LNet_TCP_AddSocket(four_player_tcp_sockets, four_player_sender[x].host_socket);
                                                                              ^
/home/rushjet1/gbeplus/5/gbe-plus-master/src/dmg/dmg07.cpp:164:7: error: ‘four_player_server’ was not declared in this scope
   if((four_player_server[x].connected) && (four_player_sender[x].connected))

I'm getting a lot of these errors.

**Comment by shonumi** [comment]
Update to the latest, and the errors should be gone.

**Comment by RushJet1** [comment]
Thanks, it compiled just fine.

Another question: Is it possible to replace tiles based on context / position in the game?  For instance, this game (and a lot of GB games) use a blank / all white tile for blank areas, but if i want to make that blue in one level and red in another level, it's not possible right now - everything will be one color.  Maybe make it depend on some memory value or something.

**Comment by shonumi** [comment]
> Another question: Is it possible to replace tiles based on context / position in the game?

Yes it's possible via two options, only one of which is currently implemented. The first is by not only using a hash of the tile, but also the address it occupies in VRAM. For example, say you have a blank tile (all-white) on the start screen and you also have a blank tile for the sky on Stage 1 of whatever game. The odds are that the game will reshuffle tiles in VRAM, so you can use the different positions in VRAM to make two different contexts. You need to select the EXT_VRAM_ADDR option when dumping a tile, and the address will automatically be added when making an entry in the manifest.

The other option (when EXT_VRAM isn't applicable, e.g. when Mega Man changes weapons) would be to associate that tile with a specific value in the Game Boy's memory. This isn't implemented, and I'm wary of adding it just yet because of the technical level required for most end-users. That is to say, only if you know how to poke around the Game Boy's WRAM (or SRAM too potentially). Additionally, there are cases where a 16-bit value is needed as opposed to an 8-bit value, which makes things a bit messier implementing it properly.

I don't think the latter option will be included any time soon. There are a lot of ways to deal with context but I'd rather not go overboard to deal with every kind of edge case.

**Comment by RushJet1** [comment]
Enabling "EXT_VRAM_ADDR" does nothing.  It must not move that tile around.  For instance, I did a metatile of one of the robot master detailed pictures that shows when you choose an enemy and changed that tile to black.  Every white space in the game is now black, from the text boxes to the title screen and the sky of every level.

**Comment by shonumi** [comment]
Yup, that's because the EXT_VRAM_ADDR only works in limited situations. If the address of the time doesn't change, it won't have the intended effect.
